### PR TITLE
8341379: Shenandoah: Improve lock contention during cleanup

### PR DESCRIPTION
**Pull request by kdnilsen**
This change improves the efficiency of cleaning up (recycling) regions that have been trashed by GC effort.  The affected code runs at the end of FinalMark to reclaim immediate garbage.  It runs at the end of FinalUpdateRefs to reclaim the regions that comprised the collection set, from which all live objects have now been evacuated.

Efficiency improvements include:
1. Rather than invoking the os (while holding the Heap lock) to obtain the time twice for every region recycled, we invoke the os only once for each batch of 32 regions that are to be processed.
2. Rather than enforcing that the loop runs no longer than 30 us, we refrain from starting a second batch of regions if more than 8 us has passed since the preceding batch was processed.

Below, each trial runs for 1 hour, processing 28,000 transactions per second.

Without this change, latency for 4 un-named business services is represented by the following chart:
image

With this change, latency for the same services is much better:
image

A comparison of the two is provided by the following:
image

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341379](https://bugs.openjdk.org/browse/JDK-8341379): Shenandoah: Improve lock contention during cleanup (**Bug** - P4)

### Reviewers
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer) Review applies to [acf517f5](https://git.openjdk.org/jdk/pull/21211/files/acf517f56ffa73d305d48d559df0fe7b9ef77ad5)
 * [Xiaolong Peng](https://openjdk.org/census#xpeng) (@pengxiaolong - Author)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21211/head:pull/21211` \
`$ git checkout pull/21211`

Update a local copy of the PR: \
`$ git checkout pull/21211` \
`$ git pull https://git.openjdk.org/jdk.git pull/21211/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21211`

View PR using the GUI difftool: \
`$ git pr show -t 21211`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21211.diff">https://git.openjdk.org/jdk/pull/21211.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21211#issuecomment-2387395739)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp [hunk 1]
```diff
@@ -920,15 +920,34 @@ void ShenandoahFreeSet::recycle_trash() {
...
+    const jlong deadline = batch_start_time + deadline_ns;
+    do {
+      // Measurements suggest it typically takes less than 200 ns on average to recycle_trash() for a single region.
+      // With an 8 us deadline and a 32-region batch, we find that we typically process 3 batches between yields.
+      //
+      // Note that deadline is enforced since the end of previous batch.  In the case that yield() or acquisition of heap lock
+      // takes a "long time", we will have less time to process regions, but we will always process at least one batch between
+      // yields.  Yielding more frequently when there is heavy contention for the heap lock or for CPU cores is considered the
+      // right thing to do.
+      const size_t REGION_STRIDE = 32;
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:944]: Maybe call this `REGIONS_PER_BATCH`? When I see `stride` I think how far the loop index moves on each iteration.

Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:944]: Good suggestion.

Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp]: Thanks.  I've made this change.

**Comment by kdnilsen** [comment]
I've got a bit more information about the differences in behavior between no-batch trial 1 and trial 2:
1. Note that trial2 has much worse p9999 latency than trial1
2. The difference is NOT safepoint behavior.  Trial 1 actually had more safepoints that lasted longer than 1 ms, with the longest lasting 5.658220ms.  The longest safepoint in trial 2 was 3.420009 ms.
3. There is evidence to suggest that the difference stems from concurrent cleanup: trial1 had 1 concurrent cleanup event taking more than 1 ms, with time of 1.142 ms, average cleanup time of 85.1 us; trial 3 had 3 concurrent cleanup events taking more than 1 ms, with the max of 1.377 ms, average cleanup time of 85.8 us.
4. For comparison, the three runs with this fix had an average concurrent cleanup event time of 69.8 us.

Qualitative assessment:

This fix allows concurrent cleanup to happen on average in 18.3% less time.  This means it is less likely to collide with a mutator thread in access to the shared heap lock.

When a collision does occur, it is resolved more quickly, allowing the the mutator to proceed in no more than 8 us plus the time to process one batch of 32 regions rather than having to wait a max of 30 us.

**Review by earthling-amzn — APPROVED** [review]
Looks good. Appreciate the comprehensive analysis.

**Comment by pengxiaolong** [comment]
Looks good to me, thanks for the detailed analysis!  

My understanding: the major benefits from this PR:
1. Minimize the cost on system call os::javaTimeNanos()
2. Not to start next batch if next it has hold the heap lock over 8us when current batch finishes.

Maybe we should calculate the next batch size based on the reminding time and speed after a batch, the assumption 200ns/region is unlikely to be true for all hardwares, in the worse case, it may hold the lock for up to 15us~16us,

**Review by pengxiaolong — APPROVED** [review]

**Comment by kdnilsen** [comment]
It's hard to "predict" the cost of the next batch.  It depends on how many of the regions within the batch might have already been recycled by on-demand actions of mutator threads.  In general, its seems the cost of a batch is typically less than 4 us, because I gathered some stats with instrumentation (no longer present) that says typical number of batches processed between yield is 3.  The first two batches, including the time to yield and acquire lock, must have completed in less than 8 us, or we would not have allowed ourself to start another batch.

True, a different hardware might take more or less time to process batch, but I don't think the behavior will be too sensitive to this.  If it takes less than 2.66 us, then we might get an average of 4 batches processed between yields.  If it takes longer than 8 us, we'll only get 1 batch processed between yields.  But we'll still be making good progress on recycling the trashed regions while remaining very responsive to mutator needs to access the heap lock.

**Comment by kdnilsen** [comment]
Thanks for reviews.

**Review by phohensee — APPROVED** [review]
I worry that we're hard-coding assumed task durations. I'm ok with this PR as it is, but I suggest we add a facility to GC initialization that does dummy tasks such as this in order to get somewhat-realistic times for use in this kind of situation.

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp [hunk 2]
```diff
@@ -920,15 +920,34 @@ void ShenandoahFreeSet::recycle_trash() {
...
-    const jlong deadline = os::javaTimeNanos() + deadline_ns;
-    while (idx < count && os::javaTimeNanos() < deadline) {
-      try_recycle_trashed(_trash_regions[idx++]);
+    if (idx > 0) {
+      os::naked_yield(); // Yield to allow allocators to take the lock, except on the first iteration
     }
+    ShenandoahHeapLocker locker(_heap->lock());
+    // Avoid another call to javaTimeNanos() if we already know time at which last batch ended
+    batch_start_time = batch_end_time;
+    const jlong deadline = batch_start_time + deadline_ns;
```
Comment (pengxiaolong) [inline src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:935]: Maybe before taking the heap lock?

Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:935]: I'll adjust this.

**Comment by kdnilsen** [comment]
I'll modify the code to adjust for current localized behavior of the host computer.  (e.g. not hard-code assumptions about task durations.)

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp [hunk 3]
```diff
@@ -919,16 +918,45 @@ void ShenandoahFreeSet::recycle_trash() {
...
+      // right thing to do.
+      const size_t REGIONS_PER_BATCH = 32;
+      size_t max_idx = MIN2(count, idx + REGIONS_PER_BATCH);
+      while (idx < max_idx) {
+        try_recycle_trashed(_trash_regions[idx++]);
+      }
+      total_batches++;
+      batch_end_time = os::javaTimeNanos();
+      // Estimate includes historic combination of yield times and heap lock acquisition times.
+      batch_process_time_estimate = (batch_end_time - recycle_trash_start_time) / total_batches;;
```
Comment (pengxiaolong) [inline src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:957]: Nit: double semicolons

**Review by pengxiaolong — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp]: Thanks.  Fixed.

**Comment by kdnilsen** [comment]
I restructured the mechanism that decides whether to start processing another batch of regions.  In the code as originally structured, we would start another batch whenever the time since previous yield is less than 8 us.  In this new version of the code, we start another batch whenever the time since previous yield plus the predicted time to process another batch is less than 10 us.  Prediction of how long a batch will take to process is based on most recent history.  Note that this is highly dependent on the state of the application.  If the application has lots of mutator threads that are urgently allocating, there will be contention for the global heap lock and there will be contention for CPU cores, and both of these will cause the batch processing time to increase.

As expected, this new version does a better job of constraining the amount of time between yields.  With this new code, we will never start processing another batch unless we have some confidence that the new batch will complete on schedule.  Previously, we might have completed processing a first batch at time 7.5 us.  Then, we would have immediately started to process a second patch, even though a prediction of batch processing time would have allowed us to predict that this second batch would not finish until time 15 us.

The new version of the code will sometimes cause mutators to wait slightly longer for yield and the heap lock when the system is "lightly loaded".  This is shown to effect p50 latencies.  When the system is lightly loaded, the time to process a batch (on current hardware) is approximately 1.5 us.   Suppose we finish processing a batch at time 8.25 us.  In the original implementation, we would have immediately yielded.  However, in this new version of the code, we'll take the next batch, because this is predicted to complete at time 9.75 us, which is less than deadline 10 us.  Making this choice allows the GC thread that is recycling trash regions to have higher throughput.

Here are the results of testing this new version of the code:

image

and this is how this new version compares to mainline without this PR:

image

Most of the p99.99 percentile latencies are considerably improved.  There is a slight degradation of p50 latencies.

Which would be considered the preferred solution?  I'll vote for the new code, but can be persuaded either way.

**Comment by pengxiaolong** [comment]
> Which would be considered the preferred solution? I'll vote for the new code, but can be persuaded either way.

New code looks good to me, the test result also looks pretty good.

**Review by pengxiaolong — APPROVED** [review]

**Review by phohensee — APPROVED** [review]

**Comment by ysramakrishna** [comment]
Nice!

/sponsor

**Pull request closed without merging**